### ISSUE TITLE
Suggestions to Barricade & Structure events

### ISSUE THREAD

**Issue by Pustalorc** [SmartlyDressedGames/Unturned-3.x-Community#1943]
The following is a change to 2 existing events, and 2 new events for both BarricadeManager and StructureManager.

TL;DR - onSpawned events should provide all data, not just model drops and regions. onDestroy events should exist and should provide the Data & Drop. Cancel-able events are not an alternative, due to their nature in the possibility of being cancelled.

`onStructureSpawned` & `onBarricadeSpawned`:
Both give us the drop and the region, but they do not provide the data.
The region does not hold any of the barricade's data beyond the drop until other post-code runs.
This means that a plugin listening to these events does not know who placed it, its health, its asset (except for BarricadeDrop, as that includes the asset, whilst StructureDrop doesn't), etc.
This also means that a plugin that is waiting for someone to place a brand new item with a specific Id to then do something is unable to, unless they wait for ~1 second in a separate thread on their code and then get the Barricade/Structure data from the provided region.
This event should provide all 3: the region, the drop and the data of the placed buildable; this way any plugin that needs to check owner after placing down a buildable, or that needs to the buildable's data in a DB, or whatever it may be, it can do it without issues.
Now, you may say "Why not use `onDeployStructureRequested` or `onDeployBarricadeRequested`?" and the answer to that is:
they can be cancelled, if my plugin loads first, and then a 2nd plugin loads after, and the 2nd plugin cancels the deployment of a barricade/structure, now the data that I collected pre-drop is useless.

`onStructureDestroyed` & `onBarricadeDestroyed` - Delegate should raise with Region & Index (if raised pre-removal from region) or with Data & Drop (if raised post-removal from region)
Removal from region: `region.drops.RemoveAt(index)` / `regions.barricades.RemoveAt(index)`
These events are needed to know when a barricade/structure gets destroyed.
`onBarricadeDestroyed` should be raised by `BarricadeManager.destroyBarricade` and `onStructureDestroyed` should be raised by `StructureManager.destroyStructure`.
You might say to use `onSalvageStructureRequested` or `onDamageStructureRequested` (or the same events but for barricades), but the same cancellation issue as above occurs.
Having a post-destroy event is simpler and can tell us all we need to know about what was destroyed.
Do note that someone subscribing to this event does **not** want to know how much damage a buildable received or if it was salvaged, they only want to know **what** was destroyed and **where** it was destroyed.
This allows for plugins that are tracking specific buildables (such as a plugin that uses signs as shops or similar) to know when their object has been destroyed and to handle it.
I have previously talked about this in #1737 

That's about the events ones that I'd be looking for.
For any other event requests/changes, I'm sure devs would love to comment here, and they might extend it to other managers as well, as there's a lot of things they'd love to have.

**Comment by SDGNelson** [comment]
No ETA yet, but I agree your proposal makes sense.
In the meantime one workaround for the destroy issue would be to add a custom component to the barricade/structure from the spawned event, and process the destruction in the OnDestroy MonoBehaviour callback.

**Comment by SDGNelson** [comment]
Closing because the issues list was not really intended for feature requests. Sorry to close this discussion - I do WANT to fulfill all (most?) requests, but it is a matter of prioritization. Constantly being reminded of all the pending requests when checking issues is stressful. I will keep the request in mind. On GitHub maybe the Ideas section of Discussions would be the best place.
